File: src/org/simbrain/custom_sims/simulations/agent_trails/ColorPlot.java
package org.simbrain.custom_sims.simulations.agent_trails;

import org.simbrain.util.projection.Halo;
import org.simbrain.workspace.updater.UpdateAction;

/**
 * Creates a halo around the predicted next point at each iteration, to make
 * clear what points are being predicted.
 */
public class ColorPlot implements UpdateAction {

    /** Reference to simulation object that has all the main variables used. */
    AgentTrails sim;

    /**
     * Construct the updater.
     */
    public ColorPlot(AgentTrails sim) {
        super();
        this.sim = sim;
    }

    @Override
    public String getDescription() {
        return "Color projection points";
    }

    @Override
    public String getLongDescription() {
        return "Color projection points";
    }

    @Override
    public void invoke() {
        double[] predictedState = sim.predictionNet.getActivations();
        Halo.makeHalo(sim.plot.getProjectionModel().getProjector(),
                predictedState, (float)sim.errorNeuron.getActivation());
    }

}

File: src/main/kotlin/org/simbrain/custom_sims/simulations/utils/ColorPlot.kt
package org.simbrain.custom_sims.simulations.utils

import org.simbrain.network.groups.NeuronGroup
import org.simbrain.util.projection.Halo
import org.simbrain.util.projection.Projector
import org.simbrain.workspace.updater.updateAction

/**
 * Creates a halo around the predicted next point at each iteration, to make
 * clear what points are being predicted.
 */
fun createColorPlotUpdateAction(projector: Projector, predictedState: DoubleArray, errorActivation: Double)
        = updateAction("Color projection points") {
    Halo.makeHalo(projector, predictedState, errorActivation.toFloat())
}

fun createColorPlotUpdateAction(projector: Projector, predictionNet: NeuronGroup, errorActivation: Double)
        = createColorPlotUpdateAction(projector, predictionNet.activations, errorActivation)